### PR TITLE
Make result of cutting up less contradicting to recipes

### LINKED ISSUE #50415: Wetsuit multiplying neoprene patches on dissassembly
### Describe the bug

Crafting a wetsuit costs 12 neoprene patches, but cutting it up gives 63 patches back

### Steps To Reproduce

1. Spawn or get enough materials to craft a wetsuit (12 neoprene patches + other stuff)
2. Drop wetsuit
3. [B]utcher it (gives 63 neoprene patches back)

### Expected behavior

Expected that disassembling a wetsuit would give me back the 12 patches, not 63.

### Screenshots

image
image

### Versions and configuration

- OS: Windows 10 Pro
- Game Version: 28c62b9 (experimental)
- Graphics version: Tiles - Chibi_UtiCa
- Ingame language: English
- Mods loaded: basegame + disable npc needs

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Menschheit**
#### Summary
Bugfixes "Rework calculation of cutting outcome"

#### Purpose of change

Fix #50415

There are two problems.
1. How many things the player can get from salvage mostly depends on weight, it may be in contradiction to the corresponding crafting recipe;
2. The codes used weight as key of a std::map. So only 1 representative for each weight value allowed, the player can never get both components (neoprene patches and rags have the same weight: 80g) from cutting a wetsuit.

#### Describe the solution

Try to decompose the item into irreducible parts, that is:
1. If an item is a basic component (neoprene patches, rags etc.), add it into the list.
2. If an item is not made of one of the materials which the original item made from, discard it.
3. If an item is not salvageable or it is count by charges, consider it as irreducible and count its weight, distribute into the materials uniformly.
4. If an item has components defined or an appropriate craft recipe(I did not take disassemble recipe into acount, since the disassembly itself may contradict to crafting recipe cuz of bad json definition) , process its components.
5. If no recipes found consider this item as irreducible, count its weight.

In the end apply item loss considering damage, skills etc (in the same proportion as before ).

#### Testing

Cutting up wetsuit will not provide more than 12 neoprene patches.

#### Additional context

Main goal of this is to make it less possible to contradict existing crafting recipes, have no Idea how it should be in real life.

There can still be weird numbers related to recipes, but ~~anyway~~ hopefully not as absurd as 63 neoprene patches vs 12. Still needs more test(I stopped finding new bugs after trying to fix old ones, for obvious reason:joy_cat:).

### CODE REVIEW

File: src/itype.cpp [hunk 1]
```diff
@@ -159,3 +159,13 @@ bool itype::can_have_charges() const
     }
     return false;
 }
+
+bool itype::is_basic_component() const
+{
+    for( const auto &mat : materials ) {
+        if( mat.obj().salvaged_into()  == cata::optional<itype_id>( get_id() ) ) {
```
Comment (anothersimulacrum) [inline src/itype.cpp:166]:
```suggestion
        if( mat->salvaged_into() && mat->salvaged_into()  == get_id() ) {
```

**Comment by kevingranade** [comment]
This seems like a reasonable approach, but please address the lgtm warning and fix the cut-up tests that are failing now.

**Comment by Nabo92** [comment]
YES! My request has been granted! Thanks for the work!

**Comment by Menschheit** [comment]
About the tests:
in iuse_actor_test.cpp, line 155
` CHECK( salvaged_mass >= ( cut_up_target_mass * 0.99 ) - smallest_yield_mass );`
It literally checks the weight, so I can't make the test happy unless I remove this line.

Another weird thing is, items like television and radio are also checked, but currently they are not even recognized as salvageable by the game cuz of this line
` if( !it.only_made_of( material_whitelist ) ) {
        return false;
    }`
in `salvage_actor::valid_to_cut_up` 

should I change `only_made_of` into "made_of_any"? Otherwise testing those items does not make any sense

**Comment by eltank** [comment]
Hi @Menschheit, could you take a look at #50944? This PR has been identified as a necessary condition for triggering the bug.

**Comment by eltank** [comment]
Never mind, figured out the bug. Calling `front()` on a vector without checking if it's empty. This is inside the `Try to find an available recipe and "restore" its components` block. Sending a PR to fix.

Note to maintainers: please don't backport this PR without also backporting the fix. It causes game crashes.

**Pull request merged**